Nondeterministic failures on travis

We've seen a few nondeterministically failing builds post 0.1.0:
- https://travis-ci.org/uber/pyro/builds/297437367
- https://travis-ci.org/uber/pyro/builds/297418157
- https://travis-ci.org/uber/pyro/builds/297417599

Two tests that I've seen fail are:
- test_gmm_elbo_gradient[dense-sample-single]
- test_gmm_elbo_gradient[flat-sample-single]

<details>

```
=================================== FAILURES ===================================
[31m[1m_________________ test_gmm_elbo_gradient[dense-sample-single] __________________[0m

model = <function gmm_model at 0x7f23f41f67b8>
guide = <function gmm_guide at 0x7f23f41f6840>, enum_discrete = False
trace_graph = False

[1m    @pytest.mark.parametrize("model,guide", [[0m
[1m        (gmm_model, gmm_guide),[0m
[1m        (gmm_batch_model, gmm_batch_guide),[0m
[1m    ], ids=["single", "batch"])[0m
[1m    @pytest.mark.parametrize("enum_discrete", [[0m
[1m        pytest.param(True,[0m
[1m                     marks=pytest.mark.xfail(reason="https://github.com/uber/pyro/issues/278")),[0m
[1m        False,[0m
[1m    ], ids=["sum", "sample"])[0m
[1m    @pytest.mark.parametrize("trace_graph", [False, True], ids=["dense", "flat"])[0m
[1m    def test_gmm_elbo_gradient(model, guide, enum_discrete, trace_graph):[0m
[1m        pyro.clear_param_store()[0m
[1m        num_particles = 4000[0m
[1m        data = Variable(torch.Tensor([-1, 1]))[0m
[1m    [0m
[1m        print("Computing gradients using surrogate loss")[0m
[1m        Elbo = TraceGraph_ELBO if trace_graph else Trace_ELBO[0m
[1m        elbo = Elbo(enum_discrete=enum_discrete,[0m
[1m                    num_particles=(1 if enum_discrete else num_particles))[0m
[1m        with xfail_if_not_implemented():[0m
[1m            elbo.loss_and_grads(model, guide, data)[0m
[1m        params = sorted(pyro.get_param_store().get_all_param_names())[0m
[1m        assert params, "no params found"[0m
[1m        actual_grads = {name: pyro.param(name).grad.clone() for name in params}[0m
[1m    [0m
[1m        print("Computing gradients using finite difference")[0m
[1m        elbo = Trace_ELBO(num_particles=num_particles)[0m
[1m        expected_grads = finite_difference(lambda: elbo.loss(model, guide, data))[0m
[1m    [0m
[1m        for name in params:[0m
[1m            print("{} {}{}{}".format(name, "-" * 30, actual_grads[name].data,[0m
[1m                                     expected_grads[name].data))[0m
[1m>       assert_equal(actual_grads, expected_grads, prec=0.5)[0m

[31m[1mtests/infer/test_enum.py[0m:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[31m[1mtests/common.py[0m:211: in assert_equal
[1m    assert_equal(x_val, y[key], prec, msg='{} {}'.format(key, msg))[0m
[31m[1mtests/common.py[0m:193: in assert_equal
[1m    assert_tensors_equal(x, y, prec, msg)[0m
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = 
-2.8286
[torch.DoubleTensor of size 1]

b = 
-3.3407
[torch.DoubleTensor of size 1]
, prec = 0.5, msg = 'p '

[1m    def assert_tensors_equal(a, b, prec=1e-5, msg=''):[0m
[1m        assert a.size() == b.size(), msg[0m
[1m        if prec == 0:[0m
[1m            assert (a == b).all(), msg[0m
[1m        elif a.numel() > 0:[0m
[1m            b = b.type_as(a)[0m
[1m            b = b.cuda(device=a.get_device()) if a.is_cuda else b.cpu()[0m
[1m            # check that NaNs are in the same locations[0m
[1m            nan_mask = a != a[0m
[1m            assert torch.equal(nan_mask, b != b), msg[0m
[1m            diff = a - b[0m
[1m            diff[nan_mask] = 0[0m
[1m            if diff.is_signed():[0m
[1m                diff = diff.abs()[0m
[1m            max_err = diff.max()[0m
[1m>           assert max_err < prec, msg[0m
[31m[1mE           AssertionError: p[0m

[31m[1mtests/common.py[0m:153: AssertionError
[31m[1m__________________ test_gmm_elbo_gradient[flat-sample-single] __________________[0m

model = <function gmm_model at 0x7f23f41f67b8>
guide = <function gmm_guide at 0x7f23f41f6840>, enum_discrete = False
trace_graph = True

[1m    @pytest.mark.parametrize("model,guide", [[0m
[1m        (gmm_model, gmm_guide),[0m
[1m        (gmm_batch_model, gmm_batch_guide),[0m
[1m    ], ids=["single", "batch"])[0m
[1m    @pytest.mark.parametrize("enum_discrete", [[0m
[1m        pytest.param(True,[0m
[1m                     marks=pytest.mark.xfail(reason="https://github.com/uber/pyro/issues/278")),[0m
[1m        False,[0m
[1m    ], ids=["sum", "sample"])[0m
[1m    @pytest.mark.parametrize("trace_graph", [False, True], ids=["dense", "flat"])[0m
[1m    def test_gmm_elbo_gradient(model, guide, enum_discrete, trace_graph):[0m
[1m        pyro.clear_param_store()[0m
[1m        num_particles = 4000[0m
[1m        data = Variable(torch.Tensor([-1, 1]))[0m
[1m    [0m
[1m        print("Computing gradients using surrogate loss")[0m
[1m        Elbo = TraceGraph_ELBO if trace_graph else Trace_ELBO[0m
[1m        elbo = Elbo(enum_discrete=enum_discrete,[0m
[1m                    num_particles=(1 if enum_discrete else num_particles))[0m
[1m        with xfail_if_not_implemented():[0m
[1m            elbo.loss_and_grads(model, guide, data)[0m
[1m        params = sorted(pyro.get_param_store().get_all_param_names())[0m
[1m        assert params, "no params found"[0m
[1m        actual_grads = {name: pyro.param(name).grad.clone() for name in params}[0m
[1m    [0m
[1m        print("Computing gradients using finite difference")[0m
[1m        elbo = Trace_ELBO(num_particles=num_particles)[0m
[1m        expected_grads = finite_difference(lambda: elbo.loss(model, guide, data))[0m
[1m    [0m
[1m        for name in params:[0m
[1m            print("{} {}{}{}".format(name, "-" * 30, actual_grads[name].data,[0m
[1m                                     expected_grads[name].data))[0m
[1m>       assert_equal(actual_grads, expected_grads, prec=0.5)[0m

[31m[1mtests/infer/test_enum.py[0m:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[31m[1mtests/common.py[0m:211: in assert_equal
[1m    assert_equal(x_val, y[key], prec, msg='{} {}'.format(key, msg))[0m
[31m[1mtests/common.py[0m:193: in assert_equal
[1m    assert_tensors_equal(x, y, prec, msg)[0m
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = 
-2.8286
[torch.DoubleTensor of size 1]

b = 
-3.3407
[torch.DoubleTensor of size 1]
, prec = 0.5, msg = 'p '

[1m    def assert_tensors_equal(a, b, prec=1e-5, msg=''):[0m
[1m        assert a.size() == b.size(), msg[0m
[1m        if prec == 0:[0m
[1m            assert (a == b).all(), msg[0m
[1m        elif a.numel() > 0:[0m
[1m            b = b.type_as(a)[0m
[1m            b = b.cuda(device=a.get_device()) if a.is_cuda else b.cpu()[0m
[1m            # check that NaNs are in the same locations[0m
[1m            nan_mask = a != a[0m
[1m            assert torch.equal(nan_mask, b != b), msg[0m
[1m            diff = a - b[0m
[1m            diff[nan_mask] = 0[0m
[1m            if diff.is_signed():[0m
[1m                diff = diff.abs()[0m
[1m            max_err = diff.max()[0m
[1m>           assert max_err < prec, msg[0m
[31m[1mE           AssertionError: p[0m

[31m[1mtests/common.py[0m:153: AssertionError
[1m============================= 46 tests deselected ==============================[0m
[31m[1m 2 failed, 446 passed, 63 skipped, 46 deselected, 13 xfailed, 2 xpassed in 2133.54 seconds [0m
Exception ignored in: <function WeakValueDictionary.__init__.<locals>.remove at 0x7f242d817d90>
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.3/lib/python3.5/weakref.py", line 117, in remove
TypeError: 'NoneType' object is not callable

travis_time:end:054b5900:start=1509849924478234476,finish=1509852063963563211,duration=2139485328735
[0K
[31;1mThe command "pytest -vs --cov=pyro --stage unit" exited with 1.[0m

Done. Your build exited with 1.
```
</details>

--- a/tests/infer/test_enum.py
+++ b/tests/infer/test_enum.py
@@ -186,7 +186,7 @@ def finite_difference(eval_loss, delta=0.1):
     params = pyro.get_param_store().get_all_param_names()
     assert params, "no params found"
     grads = {name: Variable(torch.zeros(pyro.param(name).size())) for name in params}
-    for name in params:
+    for name in sorted(params):
         value = pyro.param(name).data
         for index in itertools.product(*map(range, value.size())):
             center = value[index]

